clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java

```diff
@@ -0,0 +1,468 @@
+    }
+
+    @Override
+    public boolean hasMagic(byte magic) {
+        return magic == legacyRecord().magic();
+    }
+
+    @Override
+    public boolean hasTimestampType(TimestampType timestampType) {
+        return legacyRecord().timestampType() == timestampType;
+    }
+
+    @Override
+    public long checksum() {
+        return legacyRecord().checksum();
+    }
+
+    @Override
+    public long maxTimestamp() {
+        return timestamp();
+    }
+
+    @Override
+    public long timestamp() {
+        return legacyRecord().timestamp();
+    }
+
+    @Override
+    public TimestampType timestampType() {
+        return legacyRecord().timestampType();
+    }
+
+    @Override
+    public long baseOffset() {
+        return iterator().next().offset();
+    }
+
+    @Override
+    public byte magic() {
+        return legacyRecord().magic();
+    }
+
+    @Override
+    public CompressionType compressionType() {
+        return legacyRecord().compressionType();
+    }
+
+    @Override
+    public int sizeInBytes() {
+        return legacyRecord().sizeInBytes() + LOG_OVERHEAD;
+    }
+
+    @Override
+    public Integer countOrNull() {
+        return null;
+    }
+
+    @Override
+    public String toString() {
+        return "LegacyRecordBatch(" + offset() + ", " + legacyRecord() + ")";
+    }
+
+    @Override
+    public void writeTo(ByteBuffer buffer) {
+        writeHeader(buffer, offset(), legacyRecord().sizeInBytes());
+        buffer.put(legacyRecord().buffer().duplicate());
+    }
+
+    @Override
+    public long producerId() {
+        return RecordBatch.NO_PRODUCER_ID;
+    }
+
+    @Override
+    public short producerEpoch() {
+        return RecordBatch.NO_PRODUCER_EPOCH;
+    }
+
+    @Override
+    public long sequence() {
+        return RecordBatch.NO_SEQUENCE;
+    }
+
+    @Override
+    public int baseSequence() {
+        return RecordBatch.NO_SEQUENCE;
+    }
+
+    @Override
+    public int lastSequence() {
+        return RecordBatch.NO_SEQUENCE;
+    }
+
+    @Override
+    public boolean isTransactional() {
+        return false;
+    }
+
+    @Override
+    public int partitionLeaderEpoch() {
+        return RecordBatch.UNKNOWN_PARTITION_LEADER_EPOCH;
+    }
+
+    @Override
+    public boolean isControlRecord() {
+        return false;
+    }
+
+    /**
+     * Get an iterator for the nested entries contained within this batch. Note that
+     * if the batch is not compressed, then this method will return an iterator over the
+     * shallow record only (i.e. this object).
+     * @return An iterator over the records contained within this batch
+     */
+    @Override
+    public Iterator<Record> iterator() {
+        final Iterator<AbstractLegacyRecordBatch> iterator;
+        if (isCompressed())
+            iterator = new DeepRecordsIterator(this, false, Integer.MAX_VALUE);
+        else
+            iterator = Collections.singletonList(this).iterator();
+
+        return new AbstractIterator<Record>() {
```

Comment:
We don't need to wrap `iterator`. The following makes it compile:

```text
     @Override
     public Iterator<Record> iterator() {
-        final Iterator<AbstractLegacyRecordBatch> iterator;
+        final Iterator<Record> iterator;
         if (isCompressed())
             iterator = new DeepRecordsIterator(this, false, Integer.MAX_VALUE);
         else
-            iterator = Collections.singletonList(this).iterator();
+            iterator = Collections.<Record>singletonList(this).iterator();

-        return new AbstractIterator<Record>() {
-            @Override
-            protected Record makeNext() {
-                if (iterator.hasNext())
-                    return iterator.next();
-                return allDone();
-            }
-        };
+        return iterator;
     }

     public static void writeHeader(ByteBuffer buffer, long offset, int size) {
@@ -266,7 +259,7 @@ public abstract class AbstractLegacyRecordBatch extends AbstractRecordBatch impl
         }
     }

-    private static class DeepRecordsIterator extends AbstractIterator<AbstractLegacyRecordBatch> {
+    private static class DeepRecordsIterator extends AbstractIterator<Record> {
         private final ArrayDeque<AbstractLegacyRecordBatch> batches;
         private final long absoluteBaseOffset;
         private final byte wrapperMagic;
@@ -335,7 +328,7 @@ public abstract class AbstractLegacyRecordBatch extends AbstractRecordBatch impl
         }

         @Override
-        protected AbstractLegacyRecordBatch makeNext() {
+        protected Record makeNext() {
     public static void writeHeader(ByteBuffer buffer, long offset, int size) {
@@ -266,7 +259,7 @@ public abstract class AbstractLegacyRecordBatch extends AbstractRecordBatch impl
         }
     }

-    private static class DeepRecordsIterator extends AbstractIterator<AbstractLegacyRecordBatch> {
+    private static class DeepRecordsIterator extends AbstractIterator<Record> {
         private final ArrayDeque<AbstractLegacyRecordBatch> batches;
         private final long absoluteBaseOffset;
         private final byte wrapperMagic;
@@ -335,7 +328,7 @@ public abstract class AbstractLegacyRecordBatch extends AbstractRecordBatch impl
         }

         @Override
-        protected AbstractLegacyRecordBatch makeNext() {
+        protected Record makeNext() {
             if (batches.isEmpty())
                 return allDone();
```

What do you think?